### DISCUSSION TITLE
Save the elastic energy density at each nodal as a auxvariable and then transfer it into subapp.

### DISCUSSION THREAD

**Discussion by Yaode-Yin** [idaholab/moose#28606] [Q&A Modules: General]
Hello everyone,
I try to implement the phase field fracture model via multiapp method. In the main app, the elastic problem is solved and then I want to save the elastic energy density at each nodal in a aux variable, so that the aux variable is finally transferred into the phase field app. 
My question is that how can I save the elastic energy density calculated by  `ComputeCrackedStress` as a aux variable correctly?

**Comment by GiudGiud** [comment]
Hello

Is it a material property? You can project material properties onto auxiliary variables using auxkernels. This one for "Real" valued material properties
https://mooseframework.inl.gov/source/auxkernels/MaterialRealAux.html
there is a different auxkernel for every type of material property

Auxvariables can then be transferred. Note that you will suffer a projection error most likely between the material property and the auxiliary variable, unless the material property is actually represented exactly in the target finite element basis 

Guillaume

**Reply by Yaode-Yin** [reply]
Hi GiudGiud
Thanks for your reply. Actually I've used 'MaterialRealAux'. The code is like
```
[AuxVariables]
  [elastic_energy]
   order = FIRST
   family = MONOMIAL
  []
[]
[AuxKernels]
 [elastic_energy]
  type = MaterialRealAux
  variable = elastic_energy
  property = elastic_energy
 []
[]
```
However, this elastic_energy seems to return a volume-averaged quantity over the element. I think  what I need in a phase field fracture model is the strain energy density at each nodal or Gaussian point (if I'm not wrong). So how can I transfer them to the phase field app?

**Reply by GiudGiud** [reply]
for nodal points use a nodal variable, like a first order lagrange.

For Gaussian points, what I mentioned above applies. You project the material property to a variable, then one can evaluate that variable at the Gauss points. They will only match rigorously if the projection step was exact because the material property was actually represented exactly by the finite element family & order you picked for the auxvariable

**Reply by Yaode-Yin** [reply]
When I try to set elastic_energy as a first order lagrange variable, I receive the following errors.

`Nodal AuxKernel 'elastic_energy' attempted to reference material property 'property' 
Consider using an elemental auxiliary variable for 'elastic_energy'`
I don't know why it happens. How can I fix it?

**Reply by GiudGiud** [reply]
material properties are not defined at nodes. 
So a direct projection wont be possible. You can try to make this work
https://mooseframework.inl.gov/source/auxkernels/ProjectedMaterialPropertyNodalPatchRecoveryAux.html

otherwise you can project onto a higher order monomial first, then use the ProjectionAux to project again onto a nodal variable

There's no real reason to use a nodal variable here imo, since material properties are using on quadrature points inside the element not at nodes

**Reply by Yaode-Yin** [reply]
Yes, you're right! Thanks a lot!
But I still don't know if it is possible to transfer the positive part of the elastic strain energy density at the quadrature points. I think that's what I really need. Can I implement that in the present Transfer system?

**Reply by GiudGiud** [reply]
Transferring material properties directly, not going through an auxiliary variable, would be a lot of work
